pkg/minikube/update/update.go

```diff
@@ -49,64 +50,86 @@ const (
+	}
+	return nil
 }
 
-func MaybeUpdate(output io.Writer, githubOwner, githubRepo, binaryName, lastUpdatePath string, localVersion semver.Version) {
+func checkLatestVersion(localVersion, latestVersion semver.Version) bool {
+	if localVersion.Compare(latestVersion) < 0 {
+		return true
+	} else {
+		return false
+	}
+}
 
-	downloadBinary := binaryName + "-" + runtime.GOOS + "-" + runtime.GOARCH
-	updateLinkPrefix := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/tag/" + version.VersionPrefix
-	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+func MaybeUpdate(output io.Writer, githubOwner, githubRepo string, latestVersion semver.Version) error {
+
+	var extName string
 
-	if !shouldCheckURLVersion(lastUpdatePath) {
-		return
+	if runtime.GOOS == "windows" {
+		extName = "zip"
+	} else {
+		extName = "tgz"
 	}
-	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", latestVersion, runtime.GOOS, runtime.GOARCH, extName)
+	// updateLinkPrefix := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/tag/" + version.VersionPrefix
+	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+
+	fmt.Printf("Updating to version %s\n", latestVersion)
+	// err := updateBinary(latestVersion, archiveName, updateLinkPrefix, downloadLinkFormat)
+
+	// downloadedAchive is a string that has the path to archive on the disk
+	downloadedArchive, dir, err := downloadAndSaveArchive(latestVersion, archiveName, downloadLinkFormat)
 	if err != nil {
-		glog.Errorln(err)
-		return
+		return err
 	}
-	if localVersion.Compare(latestVersion) < 0 {
-		err := writeTimeToFile(lastUpdatePath, time.Now().UTC())
-		if err != nil {
-			fmt.Println("Failed to update last update time")
-		}
-		fmt.Fprintf(output, `A newer version of %s is available. Do you want to
-automatically update from %s%s to %s%s now? [y/N] `,
-			binaryName, version.VersionPrefix, localVersion, version.VersionPrefix, latestVersion)
 
-		var confirm string
-		fmt.Scanln(&confirm)
+	defer os.RemoveAll(dir)
 
-		if strings.ToLower(confirm) == "y" {
-			fmt.Printf("Updating to version %s\n", latestVersion)
-			updateBinary(latestVersion, downloadBinary, updateLinkPrefix, downloadLinkFormat)
-			return
-		}
+	if err := downloadAndVerifyChecksum(downloadedArchive, downloadLinkFormat, archiveName, latestVersion); err != nil {
+		return err
+	}
 
-		fmt.Println("Skipping auto-update.")
+	// Extract and replace the downloaded archive in place of current binary
+
+	currentBinary, err := osext.Executable()
+	if err != nil {
+		glog.Errorf("Cannot find current binary to execute: %s", err)
+		os.Exit(1)
 	}
-}
 
-func shouldCheckURLVersion(filePath string) bool {
-	if !viper.GetBool(config.WantUpdateNotification) {
-		return false
+	err = extractAndUpdateBinary(currentBinary, downloadedArchive, dir)
```

Comment:
This still puts extraction and update together. Why not just _extractBinary_ which in turn returns the path to the extracted binary. Then you pass this path to a _updateBinary_